src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/InstrumentationAssemblyRule.cs

```diff
@@ -14,13 +14,77 @@
 // limitations under the License.
 // </copyright>
 
+using System.Diagnostics;
+using System.Reflection;
+using System.Text.Json;
+using OpenTelemetry.AutoInstrumentation.Logging;
+
 namespace OpenTelemetry.AutoInstrumentation.RulesEngine;
 
 internal class InstrumentationAssemblyRule : Rule
 {
+    private static readonly IOtelLogger Logger = OtelLogging.GetLogger("StartupHook");
+
+    public InstrumentationAssemblyRule()
+    {
+        Name = "Instrumentation Assembly Validator";
+        Description = "Ensure that the version of the OpenTelemetry Instrumentation libraries is not older than the version used by Auto-Instrumentation.";
+    }
+
     internal override bool Evaluate()
     {
-        // TODO: implement checking logic
+        var result = true;
+
+        try
+        {
+            var ruleEngineFileLocation = Path.Combine(StartupHook.LoaderAssemblyLocation ?? string.Empty, "ruleEngine.json");
+            var ruleEngineContent = File.ReadAllText(ruleEngineFileLocation);
+            var ruleFileInfoList = JsonSerializer.Deserialize<List<RuleFileInfo>>(ruleEngineContent);
+            var entryAssembly = Assembly.GetEntryAssembly();
```

Comment:
It seems worth a comment on why checking the entry assembly is reasonable even when that doesn't guarantee that another assembly wouldn't reference the OpenTelemetry SDK (which implies that an old one was shipped with the application). The intention is that after this point auto-instrumentation will inject the OpenTelemetry SDK assemblies (the loading of instrumentation assemblies is delayed) and will fail with "could not load X.Y.Z version n.n.n" if there is a reference to earlier versions by some other assembly. Lots of to just say that the assumption here is that if there is any reference to the OpenTelemetry SDK it is very likely to be on the entry assembly.

🤔 I wonder if instead, we shouldn't catch the exceptions when trying to load the SDK and do some further probing to investigate the error. It becomes a kind of pay-for-play: if it works, it works, if it doesn't we try to get better messages on the way down.

Reply:
Sorry, forgot that the goal here is to minimize the chance of crashing the application.